blueprints/delivery.service.tls.versions.md

```diff
@@ -0,0 +1,351 @@
+quickly upgrade to a new protocol as it's released.
+
+A warning should also be issued if any Delivery Service is found to be a target
+of the Delivery Service being edited that does not support all of the TLS
+versions that the Delivery Service being edited itself does support, or if the
+Delivery Service being edited does not support all of the TLS versions supported
+by a Delivery Service of which it is a target, or if the Delivery Service being
+edited is the target of a Delivery Service which has other targets that support
+TLS versions not supported by the Delivery Service being edited.
+
+Traffic Poral should also display some kind of warning - possibly in a dialog
+box - whenever the "Restrict TLS Versions" box is checked, advising that
+clients using TLS versions not listed will become unable to retrieve the
+Delivery Service's content.
+
+Finally, Traffic Portal should display a warning whenever there are TLS versions
+enabled that exclude newer versions but specify older versions (where unknown
+versions are ignored). For example, at the time of this writing, the following
+TLS version sets should produce this warning:
+
+- 1.0
+- 1.0 and 1.1
+- 1.1
+- 1.0, 1.1, and 1.2
+- 1.1 and 1.2
+- 1.0 and 1.2
+
+### Traffic Ops Impact
+Traffic Ops will require changes to its API endpoints that return or manipulate
+Delivery Services and/or representations thereof. There are also a few database
+changes that are required, but client code should be generally unaffected,
+since it is using the the library structures that will already be changed as a
+result of changes to Traffic Ops itself.
+
+#### API Impact
+The endpoints that will need to be updated to reflect the new structure of a
+Delivery service are:
+
+- `/deliveryservices`
+- `/deliveryservices/{{ID}}`
+- `/deliveryservices/{{ID}}/safe`
+- `/deliveryservice_requests`
+- `/deliveryservice_requests/{{ID}}/assign`
+- `/deliveryservice_requests/{{ID}}/status`
+- `/servers/{{ID}}/deliveryservices`
+
+Endpoints relating to CDN Snapshots and/or monitoring configuration do not need
+changes, since those already do not represent Delivery Services in the same way
+as the rest of the API specifically because they omit data unnecessary for
+Traffic Monitor and Traffic Router, and the supported TLS versions of a
+Delivery Service do not affect monitoring or routing.
+
+As a Delivery Service's Supported TLS Versions are not considered a "safe"
+field, the only endpoints that will need changes beyond just the response
+format are:
+
+- `/deliveryservices`
+- `/deliveryservices/{{ID}}`
+- `/deliveryservice_requests`
+
+When a Delivery Service or Delivery Service Request is created or modified
+using one of these endpoints, the API MUST verify that the `tlsVersions`
+property is either `null` or an array of strings that match the regular
+expression `\d+\.\d+` (an empty array may be treated implicitly as `null` in
+request bodies, but MUST ALWAYS be emitted as `null` in responses). In the
```

Comment:
since it would never be valid to disable all TLS versions, why not just make `null` and `[]` both mean _no restrictions_?

Reply:
In requests they do. It's just more obvious in responses to use `null` IMO, because then you don't ever have to think about what it means for the collection of allowed TLS versions to be empty. `null` seems like a less "special" case than `[]`, if that makes sense.

Reply:
Oh ok. That sounds fine